File: python/simplesauce/options.py
from selenium.webdriver.chrome.options import Options as ChromeOptions
from selenium.webdriver.firefox.options import Options as FirefoxOptions
from datetime import datetime
import os
import re


class SauceOptions():

    def _set_defaults(self, browser=None):
        self.browser_name = browser if browser else 'chrome'
        self.browser_version = 'latest'
        self.platform_name = 'Windows 10'

        self.options['browserName'] = self.browser_name

    def _set_default_w3c_options(self):
        self.options['sauce:options']['browserName'] = self.browser_name
        self.options['sauce:options']['browserVersion'] = self.browser_version
        self.options['sauce:options']['platformName'] = self.platform_name
        self.options['sauce:options']['name'] = self._test_name
        self.options['sauce:options']['build'] = self._sauce_build

    def _set_default_build_name(self):
        if 'build' in self.options['sauce:options']:
            return
        # Jenkins
        elif os.environ.get("BUILD_TAG"):
            self.build = "{}: {}".format(os.environ['BUILD_NAME'], os.environ['BUILD_NUMBER'])
        # Bamboo
        elif os.environ.get('bamboo_agentId'):
            self.build = "{}: {}".format(os.environ['bamboo_shortJobName'], os.environ['bamboo_buildNumber'])
        # Travis
        elif os.environ.get('TRAVIS_JOB_ID'):
            self.build = "{}: {}".format(os.environ['TRAVIS_JOB_NAME'], os.environ['TRAVIS_JOB_NUMBER'])
        # Circle
        elif os.environ.get('CIRCLE_JOB'):
            self.build = "{}: {}".format(os.environ['CIRCLE_JOB'], os.environ['CIRCLE_BUILD_NUM'])
        # Gitlab
        elif os.environ.get('CI'):
            self.build = "{}: {}".format(os.environ['CI_JOB_NAME'], os.environ['CI_JOB_ID'])
        # Team City
        elif os.environ.get('TEAMCITY_VERSION'):
            self.build = "{}: {}".format(os.environ['TEAMCITY_PROJECT_NAME'], os.environ['BUILD_NUMBER'])
        else:
            self.build = 'Build Time: {}'.format(datetime.utcnow())

    def __init__(self, browserName=None, browserVersion=None, platformName=None, options={}):
        self.options = {}
        self.options['sauce:options'] = {}

        self._test_name = 'test'
        self._set_default_build_name()

        self._set_defaults()
        self._set_default_w3c_options()

        if not any([browserName, browserVersion, platformName, options]):
            return
        elif not any([browserName, browserVersion, platformName]):
            self.parse_options(options)
            return

        if browserName:
            self.parse_browser_name(browserName)
        if browserVersion:
            self.parse_browser_version(browserVersion)
        if platformName:
            self.parse_platform_name(platformName)

    @property
    def name(self):
        return self._test_name

    @name.setter
    def name(self, name):
        self._test_name = name

        self.options['sauce:options']['name'] = self._test_name

    @property
    def build(self):
        return self._sauce_build

    @build.setter
    def build(self, build):
        self._sauce_build = build

        self.options['sauce:options']['name'] = self._sauce_build

    def set_capability(self, key, val):
        self.options[key] = val

    def set_sauce_capability(self, key, val):
        self.options['sauce:options'][key] = val

    """TODO: create better parsing mechanisms for different capabilities."""
    def parse_browser_name(self, name):
        name = name.lower()

        if name == 'ie' or name == 'internet explorer':
            self.browser_name = 'internet explorer'
        elif name == 'edge':
            self.browser_name = 'MicrosoftEdge'
        elif name == 'safari':
            self.set_mac_os()
        else:
            self.browser_name = name

        self.options['browserName'] = self.browser_name

    def parse_browser_version(self, version):
        self.browser_version = version

        self.options['browserVersion'] = self.browser_version

    def parse_platform_name(self, platform):
        if self.browser_name.lower() == 'safari':
            self.set_mac_os()
        elif 'windows' in platform.lower():
            self.platform_name = 'Windows 10'
        else:
            self.platform_name = platform

        self.options['platformName'] = self.platform_name

    def parse_options(self, options):

        if type(options) == ChromeOptions:
            self._set_defaults()
        elif type(options) == FirefoxOptions:
            self._set_defaults('firefox')
        elif 'version' in options:
            self.browser_name = options['browserName'] if options['browserName'] else 'chrome'
            self.browser_version = options['version'] if options['version'] else 'latest'
            self.platform_name = 'Windows 10' if not options['version'] else options['version']
            self.options['browserName'] = self.browser_name
            self.options['browserVersion'] = self.browser_version
            self.options['platformName'] = self.platform_name
        elif 'sauce:options' in options:
            w3cSauceOptions = options
            self.browser_name = w3cSauceOptions.get('browserName')
            self.browser_version = w3cSauceOptions['sauce:options'].get('browserVersion')
            self.platform_name = w3cSauceOptions['sauce:options'].get('platformName')

            self.options['browserName'] = self.browser_name
            self.options['browserName'] = self.browser_name
            self.options['browserVersion'] = self.browser_version
            self.options['platformName'] = self.platform_name

            self.options['sauce:options']['name'] = w3cSauceOptions.get('name')
            self.options['sauce:options']['build'] = w3cSauceOptions.get('build')
        else:
            self.browser_name = options.get('browserName')
            self.browser_version = options.get('browserVersion')
            self.platform_name = options.get('platformName')

            self.options['browserName'] = self.browser_name
            self.options['browserVersion'] = self.browser_version
            self.options['platformName'] = self.platform_name
            self.options['sauce:options']['name'] = options.get('name')
            self.options['sauce:options']['build'] = options.get('build')

    # TODO: definitely figure this out :)
    def set_mac_os(self):
        self.browser_name = 'safari'
        self.browser_version = 'latest'
        self.platform_name = 'macOS 10.14'

        self.options['browserName'] = self.browser_name
        self.options['browserVersion'] = self.browser_version
        self.options['platformName'] = self.platform_name

File: ruby/lib/simple_sauce/options.rb
# frozen_string_literal: true

require 'selenium-webdriver'

module SimpleSauce
  class Options
    class << self
      def camel_case(str)
        str.to_s.gsub(/_([a-z])/) { Regexp.last_match(1).upcase }
      end
    end

    BROWSER_NAMES = {Selenium::WebDriver::Chrome::Options => 'chrome',
                     Selenium::WebDriver::Edge::Options => 'MicrosoftEdge',
                     Selenium::WebDriver::Firefox::Options => 'firefox',
                     Selenium::WebDriver::IE::Options => 'internet explorer',
                     Selenium::WebDriver::Safari::Options => 'safari'}.freeze

    W3C = %i[browser_name browser_version platform_name accept_insecure_certs page_load_strategy proxy
             set_window_rect timeouts unhandled_prompt_behavior strict_file_interactability].freeze

    SAUCE = %i[avoid_proxy build chromedriver_version command_timeout custom_data extended_debugging idle_timeout
               iedriver_version max_duration name parent_tunnel prerun priority public record_logs record_screenshots
               record_video screen_resolution selenium_version tags time_zone tunnel_identifier video_upload_on_pass
               capture_performance].freeze

    attr_accessor :browser_name, :browser_version, :platform_name, :accept_insecure_certs, :page_load_strategy, :proxy,
                  :set_window_rect, :timeouts, :unhandled_prompt_behavior, :strict_file_interactability, :avoid_proxy,
                  :build, :chromedriver_version, :command_timeout, :custom_data, :extended_debugging, :idle_timeout,
                  :iedriver_version, :max_duration, :name, :parent_tunnel, :prerun, :priority, :public, :record_logs,
                  :record_screenshots, :record_video, :screen_resolution, :selenium_version, :tags, :time_zone,
                  :tunnel_identifier, :video_upload_on_pass, :capture_performance
    attr_reader :selenium_options

    def initialize(**opts)
      parse_selenium_options(opts.delete(:selenium_options))
      create_variables(SAUCE + W3C, opts)
      @build ||= build_name

      @browser_name ||= selenium_options['browserName'] || 'chrome'
      @platform_name ||= 'Windows 10'
      @browser_version ||= 'latest'
    end

    def capabilities
      caps = selenium_options.dup
      W3C.each do |key|
        value = send(key)
        key = self.class.camel_case(key)
        value = parse_w3c_key(key, value)
        caps[key] = value unless value.nil?
      end
      caps['sauce:options'] = {}
      SAUCE.each do |key|
        value = send(key)
        key = self.class.camel_case(key)
        value = parse_sauce_key(key, value)
        caps['sauce:options'][key] = value unless value.nil?
      end
      caps
    end
    alias as_json capabilities

    def merge_capabilities(opts)
      opts.each do |key, value|
        raise ArgumentError, "#{key} is not a valid parameter for Options class" unless respond_to?("#{key}=")

        if value.is_a?(Hash)
          value = value.each_with_object({}) do |(old_key, val), updated|
            updated[old_key.to_sym] = val
          end
        end

        send("#{key}=", value)
      end
    end

    private

    def parse_w3c_key(key, value)
      if key == 'proxy' && value
        value.as_json
      elsif key == 'timeouts' && value
        value.each_with_object({}) do |(old_key, val), updated|
          updated[self.class.camel_case(old_key)] = val
        end
      else
        value
      end
    end

    def parse_sauce_key(key, value)
      if key == 'prerun' && value.is_a?(Hash)
        value.each_with_object({}) do |(old_key, val), updated|
          updated[self.class.camel_case(old_key)] = val
        end
      elsif key == 'customData' && value.is_a?(Hash)
        value.each_with_object({}) do |(old_key, val), updated|
          updated[self.class.camel_case(old_key)] = val
        end
      else
        value
      end
    end

    def parse_selenium_options(selenium_opts)
      opts = Array(selenium_opts)

      opts.each do |opt|
        browser = BROWSER_NAMES[opt.class]
        if browser
          @browser_name = browser
        else
          W3C.each do |capability|
            send("#{capability}=", opt[capability]) if opt[capability]
          end
        end
      end

      @selenium_options = opts.map(&:as_json).inject(:merge) || {}
    end

    def create_variables(key, opts)
      key.each do |option|
        self.class.__send__(:attr_accessor, option)
        next unless opts.key?(option)

        instance_variable_set("@#{option}", opts.delete(option))
      end
      return if opts.empty?

      raise ArgumentError, "#{opts.inspect} are not valid parameters for Options class"
    end

    def build_name
      # Jenkins
      if ENV['BUILD_TAG']
        "#{ENV['BUILD_NAME']}: #{ENV['BUILD_NUMBER']}"
      # Bamboo
      elsif ENV['bamboo_agentId']
        "#{ENV['bamboo_shortJobName']}: #{ENV['bamboo_buildNumber']}"
      # Travis
      elsif ENV['TRAVIS_JOB_ID']
        "#{ENV['TRAVIS_JOB_NAME']}: #{ENV['TRAVIS_JOB_NUMBER']}"
      # CircleCI
      elsif ENV['CIRCLE_JOB']
        "#{ENV['CIRCLE_JOB']}: #{ENV['CIRCLE_BUILD_NUM']}"
      # Gitlab
      elsif ENV['CI']
        "#{ENV['CI_JOB_NAME']}: #{ENV['CI_JOB_ID']}"
      # Team City
      elsif ENV['TEAMCITY_PROJECT_NAME']
        "#{ENV['TEAMCITY_PROJECT_NAME']}: #{ENV['BUILD_NUMBER']}"
      # Default
      else
        "Build Time - #{Time.now.to_i}"
      end
    end
  end
end
